### PR TITLE
Refactor i18n

### PR DESCRIPTION
**Pull request by gabelula**
## What does this PR do?

- Moves i18n module from framework to its own coral-i18n
- Moves translations to YAML file

## How do I test this PR?

- Run the server and check that everything is working

## To Do

- Merge translations.json into YAML files for es and en

### CODE REVIEW

File: plugins/coral-plugin-love/client/components/LoveIcon.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
```
Comment (alinhle) [inline plugins/coral-plugin-love/client/components/LoveIcon.js]: This file looks like a merge error.

File: plugins/coral-plugin-love/client/components/style.css [hunk 2]
```diff
@@ -0,0 +1,30 @@
+.love {
```
Comment (alinhle) [inline plugins/coral-plugin-love/client/components/style.css]: This file looks like a merge error.

File: plugins/coral-plugin-love/client/containers/LoveIcon.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import {compose, gql, graphql} from 'react-apollo';
```
Comment (alinhle) [inline plugins/coral-plugin-love/client/containers/LoveIcon.js]: This file looks like a merge error.

**Review by alinhle — CHANGES_REQUESTED** [review]
1. Locale data is one of the biggest contributor to bundle sizes. We don't want to include all translations of all targets and all embeds into every bundle.

2. The spanish translation for the `timeago.js` has been removed with these changes without providing an alternative solution. Ideally we want to integrate `timeago.js` into our translation framework.

File: plugins/coral-plugin-love/client/containers/LoveIcon.js (hunk 3 above)
Comment (gabelula) [reply plugins/coral-plugin-love/client/containers/LoveIcon.js]: Yes. I'm taking a look at which files came from that merge.

**Comment by gabelula** [comment]
1. Yes. I'm concern about this but not sure how it would be resolved.

2. Let me look into that. It may be also related to the merge.

File: plugins/coral-plugin-love/client/index.js [hunk 4]
```diff
@@ -1,5 +1,10 @@
 import LoveButton from './LoveButton';
 
+import translations from './translations.json';
+import {loadTranslations} from '/coral-i18n/services/i18n';
+
+Promise.all([loadTranslations(translations)]);
```
Comment (alinhle) [inline plugins/coral-plugin-love/client/index.js]: We shouldn't need a promise here, as loading translations is synchronous.

File: plugins/coral-plugin-like/client/index.js [hunk 5]
```diff
@@ -1,5 +1,10 @@
 import LikeButton from './containers/LikeButton';
 
+import translations from './translations.json';
+import {loadTranslations} from '/coral-i18n/services/i18n';
+
+Promise.all([loadTranslations(translations)]);
```
Comment (alinhle) [inline plugins/coral-plugin-like/client/index.js]: We shouldn't need a promise here, as loading translations is synchronous.

File: client/coral-i18n/services/i18n.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
+const time = require('timeago.js');
```
Comment (alinhle) [inline client/coral-i18n/services/i18n.js]: We shouldn't use a mix of nodejs require style and es6 modules, our goal is to use es6 style imports and exports for the client whenever possible.

File: client/coral-i18n/services/i18n.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+// Translations are happening at https://www.transifex.com/the-coral-project/talk-1/dashboard/.
+
+const defaultLanguage = 'en';
+
+let translations = {};
+let timeagoInstance;
+
+const fetchTranslations = (locale) => {
+  translations = require(`json-loader!yaml-loader!../locales/${locale}.yml`)[locale];
```
Comment (alinhle) [inline client/coral-i18n/services/i18n.js]: I think this has no benefit over just importing all our current locals with simple static import statements. The fetchTranslations function shouldn't be needed as all translations are already in the bundle. Looking at longterm if we want to support customising the languages to include in the bundle, than we need to do that in a webpack loader, which we can do later.

**Comment by kgardnr** [comment]
Two things to address @gabelula:

- [ ] Move the locales to the projects top level

- [ ] Avoid creating another folder `coral-i18n` in the `client` folder. I believe that each folder should have a clear role and from my understanding `coral-framework` should contain all the shared logic between all our different targets and embeds and is the correct place to put our translation service. 

@cvle let me know if I missed anything here!

**Comment by alinhle** [comment]
That looks good Kim. Plus the comments from my review.

File: plugins/coral-plugin-love/client/index.js (hunk 4 above)
Comment (gabelula) [reply plugins/coral-plugin-love/client/index.js]: Yes. I was going to remove that and then I forgot when everything was working fine :/

**Comment by gabelula** [comment]
The locales are being used by the i18n service. We move the i18n service to the top level OR we move it into the framework. They should all be in the same place.

File: plugins/coral-plugin-love/client/components/style.css (hunk 2 above)
Comment (alinhle) [reply plugins/coral-plugin-love/client/components/style.css]: This file should be removed, it doesn't exist in master (merge error).

File: plugins/coral-plugin-love/client/containers/LoveIcon.js (hunk 3 above)
Comment (alinhle) [reply plugins/coral-plugin-love/client/containers/LoveIcon.js]: This file should be removed, it doesn't exist in master (merge error).

File: plugins/coral-plugin-love/client/components/LoveIcon.js (hunk 1 above)
Comment (alinhle) [reply plugins/coral-plugin-love/client/components/LoveIcon.js]: This file should be removed, it doesn't exist in master (merge error).

File: plugins/coral-plugin-love/client/index.js [hunk 8]
```diff
@@ -1,5 +1,10 @@
 import LoveButton from './LoveButton';
 
+import translations from './translations.json';
+import {loadTranslations} from '/coral-framework/services/i18n';
+
+Promise.all([loadTranslations(translations)]);
```
Comment (alinhle) [inline plugins/coral-plugin-love/client/index.js]: This promise looks obsolete

File: client/coral-embed-stream/src/index.js [hunk 9]
```diff
@@ -12,6 +12,10 @@ import localStore, {injectReducers} from 'coral-framework/services/store';
 import AppRouter from './AppRouter';
 import {pym} from 'coral-framework';
 
+import {loadTranslations} from 'coral-framework/services/i18n';
+
+loadTranslations();
```
Comment (alinhle) [inline client/coral-embed-stream/src/index.js]: Why do we need to explicitly `loadTranslations`? All translations are already in the bundle.

File: client/coral-framework/index.js [hunk 10]
```diff
@@ -1,12 +1,13 @@
 import pym from './services/PymConnection';
-import I18n from './modules/i18n/i18n';
 import actions from './actions';
+import {loadTranslations} from '../coral-framework/services/i18n';
 
 // TODO (bc): Deprecate old actions. Spreading actions is now needed.
 
+loadTranslations();
```
Comment (alinhle) [inline client/coral-framework/index.js]: See above

File: client/coral-framework/services/i18n.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+// Translations are happening at https://www.transifex.com/the-coral-project/talk-1/dashboard/.
+
+const defaultLanguage = 'en';
+
+let translations = {};
+let timeagoInstance;
+
+const fetchTranslations = (locale) => {
+  translations = require(`json-loader!yaml-loader!../locales/${locale}.yml`)[locale];
```
Comment (alinhle) [inline client/coral-framework/services/i18n.js]: Webpack statically analyses all required files and put them in the bundle, it is not necessary to `fetchTranslations`.

File: client/coral-framework/services/i18n.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
+const time = require('timeago.js');
```
Comment (alinhle) [inline client/coral-framework/services/i18n.js]: Our goal is to use es6 modules whenever we can (in the client). All nodejs `require` code should be changed to es6 import statements.

Comment (alinhle) [reply client/coral-framework/services/i18n.js]: es6 modules are better statically analysable and enables advanced features like tree-shaking.

File: plugins/coral-plugin-like/client/index.js [hunk 13]
```diff
@@ -1,5 +1,10 @@
 import LikeButton from './containers/LikeButton';
 
+import translations from './translations.json';
+import {loadTranslations} from '/coral-framework/services/i18n';
+
+Promise.all([loadTranslations(translations)]);
```
Comment (alinhle) [inline plugins/coral-plugin-like/client/index.js]: Another Promise here.

**Comment by alinhle** [comment]
image
Some more missing keys. This needs thorough testing, to catch all missing keys.

File: client/coral-admin/src/containers/ModerationQueue/helpers/moderationQueueActionsMap.js [hunk 14]
```diff
@@ -5,10 +5,10 @@ export const actionsMap = {
...
 export const menuActionsMap = {
-  'REJECT': {status: 'REJECTED', text: 'Reject', icon: 'close', key: 'r'},
-  'REJECTED': {status: 'REJECTED', text: 'Rejected', icon: 'close'},
-  'APPROVE': {status: 'ACCEPTED', text: 'Approve', icon: 'done', key: 't'},
-  'FLAGGED': {status: 'FLAGGED', text: 'Flag', icon: 'flag', filter: 'Untouched'},
-  'BAN': {status: 'BANNED', text: 'Ban User', icon: 'not interested'},
+  'REJECT': {status: 'REJECTED', text: 'reject', icon: 'close', key: 'r'},
+  'REJECTED': {status: 'REJECTED', text: 'rejected', icon: 'close'},
+  'APPROVE': {status: 'ACCEPTED', text: 'approve', icon: 'done', key: 't'},
+  'FLAGGED': {status: 'FLAGGED', text: 'rlag', icon: 'flag', filter: 'Untouched'},
```
Comment (alinhle) [inline client/coral-admin/src/containers/ModerationQueue/helpers/moderationQueueActionsMap.js]: I think this should be `flag`, also there seems to be no translation. (is a translation necessary?) ...

File: client/coral-admin/src/containers/ModerationQueue/helpers/moderationQueueActionsMap.js [hunk 15]
```diff
@@ -5,10 +5,10 @@ export const actionsMap = {
...
+  'BAN': {status: 'BANNED', text: 'ban_user', icon: 'not interested'},
```
Comment (alinhle) [inline client/coral-admin/src/containers/ModerationQueue/helpers/moderationQueueActionsMap.js]: Do we need a translation for modequeue.ban_user?

**Comment by alinhle** [comment]
image
Still some missing keys

File: client/coral-admin/src/containers/ModerationQueue/helpers/moderationQueueActionsMap.js (hunk 15 above)
Comment (gabelula) [reply client/coral-admin/src/containers/ModerationQueue/helpers/moderationQueueActionsMap.js]: Done

**Comment by gabelula** [comment]
Added translation keys.

**Comment by alinhle** [comment]
Did you forget to push?

**Comment by alinhle** [comment]
I still have these warnings:
image

**Comment by gabelula** [comment]
Can you pull again? That commit is already pushed with all those strings.

**Comment by alinhle** [comment]
I am testing on the newest commit. They appear when I go to the Community view in admin.

**Comment by alinhle** [comment]
Still a bunch of missing keys
screen shot 2017-05-25 at 19 56 38
screen shot 2017-05-25 at 19 59 50
screen shot 2017-05-25 at 19 55 54
screen shot 2017-05-25 at 20 00 11
image

**Comment by alinhle** [comment]
Some of the above keys are still missing..
image
image
image

**Review by alinhle — APPROVED** [review]

**Comment by alinhle** [comment]
Approving everything but my own commits as usual :-)

**Pull request merged**